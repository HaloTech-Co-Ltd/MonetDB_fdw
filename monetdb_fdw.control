# monetdb_fdw extension
comment = 'foreign-data wrapper for remote MonetDB servers'
default_version = '1.0'
module_pathname = '$libdir/monetdb_fdw'
relocatable = true
