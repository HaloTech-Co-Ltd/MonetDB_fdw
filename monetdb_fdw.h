/*-------------------------------------------------------------------------
 *
 * monetdb_fdw.h
 *		  Foreign-data wrapper for remote MonetDB databases
 *
 * Portions Copyright (c) 2025, Halo Tech Co.,Ltd. All rights reserved.
 * Portions Copyright (c) 2012-2023, PostgreSQL Global Development Group
 * 
 * Author: zengman <zengman@halodbtech.com>
 * 
 * IDENTIFICATION
 *		  monetdb_fdw.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef MONETDB_FDW_H
#define MONETDB_FDW_H

#include <mapi.h>

#include "foreign/foreign.h"
#include "lib/stringinfo.h"
#include "nodes/execnodes.h"
#include "nodes/pathnodes.h"
#include "utils/relcache.h"

/* like oracle_fdw */
#define OPT_KEY "key"

#define MAPI_AUTO	0	/* automatic type detection */
#define MAPI_TINY	1
#define MAPI_UTINY	2
#define MAPI_SHORT	3
#define MAPI_USHORT	4
#define MAPI_INT	5
#define MAPI_UINT	6
#define MAPI_LONG	7
#define MAPI_ULONG	8
#define MAPI_LONGLONG	9
#define MAPI_ULONGLONG	10
#define MAPI_CHAR	11
#define MAPI_VARCHAR	12
#define MAPI_FLOAT	13
#define MAPI_DOUBLE	14
#define MAPI_DATE	15
#define MAPI_TIME	16
#define MAPI_DATETIME	17
#define MAPI_NUMERIC	18

#define die(dbh,hdl)																		\
			do {																			\
				if (hdl)																	\
					elog(ERROR, "[MonetDB RESULT ERROR] %s", mapi_result_error(hdl)); 		\
				else if (dbh)																\
					elog(ERROR, "[MonetDB ERROR] %s", mapi_error_str(dbh));					\
				else																		\
					elog(ERROR, "command failed\n"); 										\
			} while (0)

#define error_info(dbh,hdl)																		\
			do {																			\
				if (hdl)																	\
					elog(INFO, "[MonetDB RESULT ERROR] %s", mapi_result_error(hdl)); 		\
				else if (dbh)																\
					elog(INFO, "[MonetDB ERROR] %s", mapi_error_str(dbh));					\
			} while (0)
/*
 * FDW-specific planner information kept in RelOptInfo.fdw_private for a
 * monetdb_fdw foreign table.  For a baserel, this struct is created by
 * MonetDB_GetForeignRelSize, although some fields are not filled till later.
 * MonetDB_GetForeignJoinPaths creates it for a joinrel, and
 * MonetDB_GetForeignUpperPaths creates it for an upperrel.
 */
typedef struct MonetdbFdwRelationInfo
{
	/*
	 * True means that the relation can be pushed down. Always true for simple
	 * foreign scan.
	 */
	bool		pushdown_safe;

	/*
	 * Restriction clauses, divided into safe and unsafe to pushdown subsets.
	 * All entries in these lists should have RestrictInfo wrappers; that
	 * improves efficiency of selectivity and cost estimation.
	 */
	List	   *remote_conds;
	List	   *local_conds;

	/* Actual remote restriction clauses for scan (sans RestrictInfos) */
	List	   *final_remote_exprs;

	/* Bitmap of attr numbers we need to fetch from the remote server. */
	Bitmapset  *attrs_used;

	/* True means that the query_pathkeys is safe to push down */
	bool		qp_is_pushdown_safe;

	/* Cost and selectivity of local_conds. */
	QualCost	local_conds_cost;
	Selectivity local_conds_sel;

	/* Selectivity of join conditions */
	Selectivity joinclause_sel;

	/* Estimated size and cost for a scan, join, or grouping/aggregation. */
	double		rows;
	int			width;
	Cost		startup_cost;
	Cost		total_cost;

	/*
	 * Estimated number of rows fetched from the foreign server, and costs
	 * excluding costs for transferring those rows from the foreign server.
	 * These are only used by estimate_path_cost_size().
	 */
	double		retrieved_rows;
	Cost		rel_startup_cost;
	Cost		rel_total_cost;

	/* Options extracted from catalogs. */
	bool		use_remote_estimate;
	Cost		fdw_startup_cost;
	Cost		fdw_tuple_cost;
	List	   *shippable_extensions;	/* OIDs of shippable extensions */
	bool		async_capable;

	/* Cached catalog information. */
	ForeignTable *table;
	ForeignServer *server;
	UserMapping *user;			/* only set in use_remote_estimate mode */

	int			fetch_size;		/* fetch size for this remote table */

	/*
	 * Name of the relation, for use while EXPLAINing ForeignScan.  It is used
	 * for join and upper relations but is set for all relations.  For a base
	 * relation, this is really just the RT index as a string; we convert that
	 * while producing EXPLAIN output.  For join and upper relations, the name
	 * indicates which base foreign tables are included and the join type or
	 * aggregation type used.
	 */
	char	   *relation_name;

	/* Join information */
	RelOptInfo *outerrel;
	RelOptInfo *innerrel;
	JoinType	jointype;
	/* joinclauses contains only JOIN/ON conditions for an outer join */
	List	   *joinclauses;	/* List of RestrictInfo */

	/* Upper relation information */
	UpperRelationKind stage;

	/* Grouping information */
	List	   *grouped_tlist;

	/* Subquery information */
	bool		make_outerrel_subquery; /* do we deparse outerrel as a
										 * subquery? */
	bool		make_innerrel_subquery; /* do we deparse innerrel as a
										 * subquery? */
	Relids		lower_subquery_rels;	/* all relids appearing in lower
										 * subqueries */

	/*
	 * Index of the relation.  It is used to create an alias to a subquery
	 * representing the relation.
	 */
	int			relation_index;
} MonetdbFdwRelationInfo;

/* in connection.c */
extern Mapi GetConnection(UserMapping *user, ForeignServer *server);
extern void ReleaseConnection(Mapi conn);
extern void do_sql_command(Mapi conn, const char *sql);

/* in monetdb_fdw.c */
extern int	set_transmission_modes(void);
extern void reset_transmission_modes(int nestlevel);
extern bool getBoolVal(DefElem *def);

/* in deparse.c */
extern void classifyConditions(PlannerInfo *root,
							   RelOptInfo *baserel,
							   List *input_conds,
							   List **remote_conds,
							   List **local_conds);
extern bool is_foreign_expr(PlannerInfo *root,
							RelOptInfo *baserel,
							Expr *expr);
extern bool is_foreign_param(PlannerInfo *root,
							 RelOptInfo *baserel,
							 Expr *expr);
extern bool is_foreign_pathkey(PlannerInfo *root,
							   RelOptInfo *baserel,
							   PathKey *pathkey);
extern void deparseInsertSql(StringInfo buf, RangeTblEntry *rte,
							 Index rtindex, Relation rel,
							 List *targetAttrs, bool doNothing,
							 List *withCheckOptionList, List *returningList,
							 List **retrieved_attrs, int *values_end_len);
extern void rebuildInsertSql(StringInfo buf, Relation rel,
							 char *orig_query, List *target_attrs,
							 int values_end_len, int num_params,
							 int num_rows);
extern void deparseUpdateSql(StringInfo buf, RangeTblEntry *rte,
							 Index rtindex, Relation rel,
							 List *targetAttrs,
							 List *withCheckOptionList, List *returningList,
							 List **retrieved_attrs);
extern void deparseDirectUpdateSql(StringInfo buf, PlannerInfo *root,
								   Index rtindex, Relation rel,
								   RelOptInfo *foreignrel,
								   List *targetlist,
								   List *targetAttrs,
								   List *remote_conds,
								   List **params_list,
								   List *returningList,
								   List **retrieved_attrs);
extern void deparseDeleteSql(StringInfo buf, RangeTblEntry *rte,
							 Index rtindex, Relation rel,
							 List *returningList,
							 List **retrieved_attrs);
extern void deparseDirectDeleteSql(StringInfo buf, PlannerInfo *root,
								   Index rtindex, Relation rel,
								   RelOptInfo *foreignrel,
								   List *remote_conds,
								   List **params_list,
								   List *returningList,
								   List **retrieved_attrs);
extern void deparseAnalyzeSizeSql(StringInfo buf, Relation rel);
extern void deparseAnalyzeInfoSql(StringInfo buf, Relation rel);
extern void deparseTruncateSql(StringInfo buf,
							   List *rels,
							   DropBehavior behavior,
							   bool restart_seqs);
extern void deparseStringLiteral(StringInfo buf, const char *val);
extern EquivalenceMember *find_em_for_rel(PlannerInfo *root,
										  EquivalenceClass *ec,
										  RelOptInfo *rel);
extern EquivalenceMember *find_em_for_rel_target(PlannerInfo *root,
												 EquivalenceClass *ec,
												 RelOptInfo *rel);
extern List *build_tlist_to_deparse(RelOptInfo *foreignrel);
extern void deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root,
									RelOptInfo *rel, List *tlist,
									List *remote_conds, List *pathkeys,
									bool has_final_sort, bool has_limit,
									bool is_subquery,
									List **retrieved_attrs, List **params_list);
extern const char *get_jointype_name(JoinType jointype);

/* in shippable.c */
extern bool is_builtin(Oid objectId);
extern bool is_shippable(Oid objectId, Oid classId, MonetdbFdwRelationInfo *fpinfo);

#endif							/* MONETDB_FDW_H */
