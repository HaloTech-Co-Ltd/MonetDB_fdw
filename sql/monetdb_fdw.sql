-- !!! If you want to run regression tests, you need to create a database named "test" in MonetDB with the default port 50000.

CREATE EXTENSION monetdb_fdw;

CREATE SERVER foreign_server FOREIGN DATA WRAPPER monetdb_fdw
OPTIONS (host '127.0.0.1', port '50000', dbname 'test');

CREATE USER MAPPING FOR CURRENT_USER SERVER foreign_server OPTIONS (user 'monetdb', password 'monetdb');

SELECT monetdb_execute('foreign_server', $$DROP USER test_u$$);
SELECT monetdb_execute('foreign_server', $$CREATE USER "test_u" WITH PASSWORD 'test_u' NAME 'test_user' SCHEMA "sys"$$);
SELECT monetdb_execute('foreign_server', $$CREATE SCHEMA IF NOT EXISTS "test_u" AUTHORIZATION "test_u"$$);
SELECT monetdb_execute('foreign_server', $$ALTER USER "test_u" SET SCHEMA "test_u"$$);

DROP USER MAPPING FOR CURRENT_USER SERVER foreign_server;
CREATE USER MAPPING FOR CURRENT_USER SERVER foreign_server OPTIONS (user 'test_u', password 'test_u');

SELECT monetdb_execute('foreign_server', $$CREATE TABLE emp( name VARCHAR(20), age INTEGER)$$);

CREATE FOREIGN TABLE emp(
        name VARCHAR(20),
        age INTEGER
)
SERVER foreign_server
OPTIONS (schema_name 'test_u', table_name 'emp');

\des
\d+ emp

-- test insert 
INSERT INTO emp VALUES('John', 23);
INSERT INTO emp VALUES('Mary', 22);

-- test selectt
SELECT * FROM emp;
SELECT monetdb_execute('foreign_server', $$SELECT * FROM emp$$);

-- test explain
-- EXPLAIN (COSTS OFF) SELECT * FROM emp;
EXPLAIN (COSTS OFF) INSERT INTO emp VALUES('Mary test', 21);

-- test truncate
TRUNCATE emp;
SELECT * FROM emp;
INSERT INTO emp VALUES('John', 23);
INSERT INTO emp VALUES('Mary', 22);
TRUNCATE TABLE emp;
SELECT * FROM emp;

SELECT monetdb_execute('foreign_server', $$DROP TABLE emp$$);
SELECT monetdb_execute('foreign_server', $$ALTER USER test_u SET SCHEMA sys$$);
SELECT monetdb_execute('foreign_server', $$DROP SCHEMA test_u$$);
SELECT monetdb_execute('foreign_server', $$DROP USER test_u$$);
