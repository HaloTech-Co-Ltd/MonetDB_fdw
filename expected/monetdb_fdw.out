-- !!! If you want to run regression tests, you need to create a database named "test" in MonetDB with the default port 50000.
CREATE EXTENSION monetdb_fdw;
CREATE SERVER foreign_server FOREIGN DATA WRAPPER monetdb_fdw
OPTIONS (host '127.0.0.1', port '50000', dbname 'test');
CREATE USER MAPPING FOR CURRENT_USER SERVER foreign_server OPTIONS (user 'monetdb', password 'monetdb');
SELECT monetdb_execute('foreign_server', $$DROP USER test_u$$);
ERROR:  [MonetDB ERROR] DROP USER: no such user role 'test_u'

SELECT monetdb_execute('foreign_server', $$CREATE USER "test_u" WITH PASSWORD 'test_u' NAME 'test_user' SCHEMA "sys"$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$CREATE SCHEMA IF NOT EXISTS "test_u" AUTHORIZATION "test_u"$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$ALTER USER "test_u" SET SCHEMA "test_u"$$);
 monetdb_execute 
-----------------
 
(1 row)

DROP USER MAPPING FOR CURRENT_USER SERVER foreign_server;
CREATE USER MAPPING FOR CURRENT_USER SERVER foreign_server OPTIONS (user 'test_u', password 'test_u');
SELECT monetdb_execute('foreign_server', $$CREATE TABLE emp( name VARCHAR(20), age INTEGER)$$);
 monetdb_execute 
-----------------
 
(1 row)

CREATE FOREIGN TABLE emp(
        name VARCHAR(20),
        age INTEGER
)
SERVER foreign_server
OPTIONS (schema_name 'test_u', table_name 'emp');
\des
             List of foreign servers
      Name      |  Owner   | Foreign-data wrapper 
----------------+----------+----------------------
 foreign_server | postgres | monetdb_fdw
(1 row)

\d+ emp
                                              Foreign table "public.emp"
 Column |         Type          | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------------------+-----------+----------+---------+-------------+----------+--------------+-------------
 name   | character varying(20) |           |          |         |             | extended |              | 
 age    | integer               |           |          |         |             | plain    |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'emp')

-- test insert 
INSERT INTO emp VALUES('John', 23);
INSERT INTO emp VALUES('Mary', 22);
-- test select
SELECT * FROM emp;
 name | age 
------+-----
 John |  23
 Mary |  22
(2 rows)

SELECT monetdb_execute('foreign_server', $$SELECT * FROM emp$$);
INFO:  
name,age
[ "John",	23	]
[ "Mary",	22	]

 monetdb_execute 
-----------------
 
(1 row)

-- test explain
-- EXPLAIN (COSTS OFF) SELECT * FROM emp;
EXPLAIN (COSTS OFF) INSERT INTO emp VALUES('Mary test', 21);
                              QUERY PLAN                              
----------------------------------------------------------------------
 Insert on emp
   MonetDB statement: INSERT INTO test_u.emp(name, age) VALUES (?, ?)
   ->  Result
(3 rows)

-- test truncate
TRUNCATE emp;
SELECT * FROM emp;
 name | age 
------+-----
(0 rows)

INSERT INTO emp VALUES('John', 23);
INSERT INTO emp VALUES('Mary', 22);
TRUNCATE TABLE emp;
SELECT * FROM emp;
 name | age 
------+-----
(0 rows)

SELECT monetdb_execute('foreign_server', $$CREATE TABLE delete_emp(name VARCHAR(20) PRIMARY KEY, age INTEGER)$$);
 monetdb_execute 
-----------------
 
(1 row)

CREATE FOREIGN TABLE delete_emp(
        name VARCHAR(20),
        age INTEGER
)
SERVER foreign_server
OPTIONS (schema_name 'test_u', table_name 'delete_emp');
INSERT INTO delete_emp VALUES('John', 23);
INSERT INTO delete_emp VALUES('Mary', 22);
SELECT * FROM delete_emp;
 name | age 
------+-----
 John |  23
 Mary |  22
(2 rows)

INSERT INTO delete_emp VALUES('Mary', 22); -- error
ERROR:  [MonetDB ERROR] INSERT INTO: PRIMARY KEY constraint 'delete_emp.delete_emp_name_pkey' violated

INSERT INTO delete_emp VALUES('Mary2', 22);
SELECT * FROM delete_emp;
 name  | age 
-------+-----
 John  |  23
 Mary  |  22
 Mary2 |  22
(3 rows)

-- test delete
DELETE FROM delete_emp WHERE name = 'John'; -- error, need set key
ERROR:  no primary key column specified for foreign MonetDB table
DETAIL:  For UPDATE or DELETE, at least one foreign table column must be marked as primary key column.
ALTER FOREIGN TABLE delete_emp ALTER name OPTIONS (ADD key 'true');
DELETE FROM delete_emp WHERE name = 'John';
SELECT * FROM delete_emp;
 name  | age 
-------+-----
 Mary  |  22
 Mary2 |  22
(2 rows)

DELETE FROM delete_emp WHERE age = 22;
SELECT * FROM delete_emp;
 name | age 
------+-----
(0 rows)

set client_min_messages = 'debug2';
INSERT INTO emp VALUES('John', 23);
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote prepare query is: INSERT INTO test_u.emp(name, age) VALUES (?, ?)
DEBUG:  monetdb_fdw bind value0: John
DEBUG:  monetdb_fdw bind value1: 23
INSERT INTO emp VALUES('Mary', 22);
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote prepare query is: INSERT INTO test_u.emp(name, age) VALUES (?, ?)
DEBUG:  monetdb_fdw bind value0: Mary
DEBUG:  monetdb_fdw bind value1: 22
SELECT * FROM emp;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.emp
 name | age 
------+-----
 John |  23
 Mary |  22
(2 rows)

INSERT INTO delete_emp VALUES('John', 23);
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote prepare query is: INSERT INTO test_u.delete_emp(name, age) VALUES (?, ?)
DEBUG:  monetdb_fdw bind value0: John
DEBUG:  monetdb_fdw bind value1: 23
INSERT INTO delete_emp VALUES('Mary', 22);
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote prepare query is: INSERT INTO test_u.delete_emp(name, age) VALUES (?, ?)
DEBUG:  monetdb_fdw bind value0: Mary
DEBUG:  monetdb_fdw bind value1: 22
SELECT * FROM delete_emp;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp
 name | age 
------+-----
 John |  23
 Mary |  22
(2 rows)

-- test update
UPDATE emp SET name = 'Mary2' WHERE name = 'Mary'; -- error
ERROR:  no primary key column specified for foreign MonetDB table
DETAIL:  For UPDATE or DELETE, at least one foreign table column must be marked as primary key column.
UPDATE delete_emp SET name = 'Mary2' WHERE name = 'Mary'; -- ok
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp WHERE ((name = 'Mary')) /* FOR UPDATE */
DEBUG:  monetdb_fdw remote prepare query is: UPDATE test_u.delete_emp SET name = ? WHERE name = ?
DEBUG:  monetdb_fdw bind value0: Mary2
DEBUG:  monetdb_fdw bind value1: Mary
SELECT * FROM delete_emp;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp
 name  | age 
-------+-----
 John  |  23
 Mary2 |  22
(2 rows)

UPDATE delete_emp SET name = 'John2' WHERE age = 23; 
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp WHERE ((age = 23)) /* FOR UPDATE */
DEBUG:  monetdb_fdw remote prepare query is: UPDATE test_u.delete_emp SET name = ? WHERE name = ?
DEBUG:  monetdb_fdw bind value0: John2
DEBUG:  monetdb_fdw bind value1: John
SELECT * FROM delete_emp;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp
 name  | age 
-------+-----
 John2 |  23
 Mary2 |  22
(2 rows)

-- test returning
INSERT INTO delete_emp VALUES('John', 23) RETURNING *;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote prepare query is: INSERT INTO test_u.delete_emp(name, age) VALUES (?, ?) RETURNING name, age
DEBUG:  monetdb_fdw bind value0: John
DEBUG:  monetdb_fdw bind value1: 23
 name | age 
------+-----
 John |  23
(1 row)

DELETE FROM delete_emp WHERE name = 'John' RETURNING *;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name FROM test_u.delete_emp WHERE ((name = 'John')) /* FOR UPDATE */
DEBUG:  monetdb_fdw remote prepare query is: DELETE FROM test_u.delete_emp WHERE name = ? RETURNING name, age
DEBUG:  monetdb_fdw bind value0: John
 name | age 
------+-----
 John |  23
(1 row)

SELECT * FROM delete_emp;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp
 name  | age 
-------+-----
 John2 |  23
 Mary2 |  22
(2 rows)

UPDATE delete_emp SET name = 'Mary' WHERE name = 'Mary2' RETURNING *; -- error, MonetDB "UPDATE ... RETURNING" has bug
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp WHERE ((name = 'Mary2')) /* FOR UPDATE */
DEBUG:  monetdb_fdw remote prepare query is: UPDATE test_u.delete_emp SET name = ? WHERE name = ? RETURNING name, age
DEBUG:  monetdb_fdw bind value0: Mary
DEBUG:  monetdb_fdw bind value1: Mary2
ERROR:  [MonetDB ERROR] (null)
SELECT * FROM delete_emp;
DEBUG:  monetdb: host: 127.0.0.1 port: 50000 user: test_u password: test_u dbname: test
DEBUG:  monetdb_fdw remote query is: SELECT name, age FROM test_u.delete_emp
 name  | age 
-------+-----
 John2 |  23
 Mary2 |  22
(2 rows)

set client_min_messages = 'INFO';
SELECT monetdb_execute('foreign_server', $$DROP TABLE delete_emp$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$DROP TABLE emp$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$ALTER USER test_u SET SCHEMA sys$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$DROP SCHEMA test_u$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$DROP USER test_u$$);
 monetdb_execute 
-----------------
 
(1 row)

